### ISSUE TITLE
twisted.web2.http_headers.parseWWWAuthenticate doesn't work without a challenge.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11206]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)| @dreid reported|
|-|-|
|Trac ID|trac#1723|
|Type|defect|
|Created|2006-05-18 18:19:18Z|

Not all HTTP Auth schemes send a challenge in the WWW-Authenticate header, NTLM for example starts it's handshake with a single WWW-Authenticate header with a value of 'NTLM'.  The following interactive session demonstrates this problem.

```
>>> from twisted.web2.http_headers import Headers
>>> h = Headers()
>>> h.setHeader('www-authenticate', [('NTLM', {})])
>>> h.getRawHeaders('www-authenticate')        
['NTLM ']
>>> h = Headers()
>>> h.setRawHeaders('www-authenticate', ['NTLM '])
>>> h.getHeader('www-authenticate')
[(None, {'ntlm': None})]
>>> h = Headers()
>>> h.setRawHeaders('www-authenticate', ['NTLM'])
>>> h.getHeader('www-authenticate')
[(None, {'ntlm': None})]
```

Attachments:

* [1723.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2d2/2d201417bd4c52bb8d2fd97eb9e44cbaaf8bff67/d44ec721c266cd7aeed5cd28675fe4b8867df537.diff) (1433 bytes) - added by dreid on 2006-05-18 18:51:39Z - Fix and Unit-Test
* [1723-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2d2/2d201417bd4c52bb8d2fd97eb9e44cbaaf8bff67/a4df1b298632d3edd1645bf6bc27f4ea131f5fc4.diff) (2037 bytes) - added by dreid on 2006-05-18 19:21:19Z - Fix typo in comment, and extend comments and test to note lack of support for comma seperated challengeless headers.

<details><summary>Searchable metadata</summary>

```
trac-id__1723 1723
type__defect defect
reporter__dreid dreid
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web2 web2
keywords__ 
time__1147976358000000 1147976358000000
changetime__1149123170000000 1149123170000000
version__ 
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to @jyknight|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to @dreid|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to @radix|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to @wsanchez|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set owner to @dreid|
|-|-|

Makes sense.  Go fer it.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid commented|
|-|-|

branched to [branches/parseWWWAuthenticate-1723](branches/parseWWWAuthenticate-1723)

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set status to `closed`|
|-|-|

(In [16980]) Improve WWW-Authenticate header parsing to include support for scheme[ kvpairs], scheme[ base64] So that headers for Negotiate and NTLM are properly parsed even when condensed to single line headers with any combination of schemes.

Author: dreid
Reviewer: wsanchez
Fixes [#11206](https://github.com/twisted/twisted/issues/11206)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
